File: databases/leveldb/python/src/recipe_kv_store.py
"""
Recipe: Basic KV Store Operations
Database: LevelDB
Description: Demonstrates opening a database, putting, getting, deleting
    keys, and performing an iterator scan.

Usage: python src/recipe_kv_store.py
"""

import os
import shutil
import plyvel


def main():
    db_path = "/tmp/leveldb_recipe_kv_store"

    # 1. Setup -- open the database (creates if missing)
    db = plyvel.DB(db_path, create_if_missing=True)

    # 2. Write -- insert key-value pairs
    db.put(b"name", b"LevelDB")
    db.put(b"type", b"key-value")
    db.put(b"year", b"2011")
    db.put(b"license", b"BSD-3-Clause")

    # 3. Read -- retrieve individual keys
    name = db.get(b"name")
    print(f"name = {name.decode()}")

    db_type = db.get(b"type")
    print(f"type = {db_type.decode()}")

    # 4. Delete -- remove a key
    db.delete(b"license")
    deleted = db.get(b"license")
    print(f"license after delete = {deleted}")

    # 5. Scan -- iterate over all entries (sorted by key)
    print("\nAll entries:")
    for key, value in db:
        print(f"  {key.decode()} => {value.decode()}")

    # 6. Cleanup
    db.close()
    shutil.rmtree(db_path, ignore_errors=True)
    print("\nDone.")


if __name__ == "__main__":
    main()

File: databases/fjall/rust/src/bin/recipe_kv_store.rs
//! Recipe: Basic KV Store Operations
//! Database: fjall
//! Description: Demonstrates opening a database, inserting, getting, deleting
//!     keys, and performing prefix scans and range iteration using the fjall
//!     keyspace API.
//!
//! Usage: cargo run --bin recipe_kv_store

use fjall::{Database, KeyspaceCreateOptions, PersistMode};

fn main() -> fjall::Result<()> {
    let dir = tempfile::tempdir()?;
    let db_path = dir.path().join("demo");

    // 1. Setup -- create the database
    let db = Database::builder(&db_path).open()?;
    let items = db.keyspace("kv", || KeyspaceCreateOptions::default())?;

    // 2. Write -- insert key-value pairs
    items.insert("name", b"fjall")?;
    items.insert("type", b"key-value")?;
    items.insert("year", b"2024")?;
    items.insert("license", b"Apache-2.0")?;

    // 3. Read -- retrieve individual keys
    match items.get("name")? {
        Some(bytes) => println!("name = {}", String::from_utf8_lossy(&bytes)),
        None => println!("name = not found"),
    }
    match items.get("type")? {
        Some(bytes) => println!("type = {}", String::from_utf8_lossy(&bytes)),
        None => println!("type = not found"),
    }

    // 4. Delete -- remove a key
    items.remove("license")?;

    match items.get("license")? {
        Some(bytes) => println!("license after delete = {}", String::from_utf8_lossy(&bytes)),
        None => println!("license after delete = None"),
    }

    // 5. Prefix scan -- iterate over entries with a common prefix
    let prefix_items = db.keyspace("prefixed", || KeyspaceCreateOptions::default())?;
    prefix_items.insert("user:1:name", b"Alice")?;
    prefix_items.insert("user:1:email", b"alice@example.com")?;
    prefix_items.insert("user:2:name", b"Bob")?;
    prefix_items.insert("user:2:email", b"bob@example.com")?;

    println!("\nAll entries for user:1:");
    for guard in prefix_items.prefix("user:1") {
        let (key, value) = guard.into_inner()?;
        println!(
            "  {} => {}",
            String::from_utf8_lossy(&key),
            String::from_utf8_lossy(&value)
        );
    }

    // 6. Range iteration
    println!("\nAll entries in range [user, user~):");
    for guard in prefix_items.range("user"..="user~") {
        let (key, value) = guard.into_inner()?;
        println!(
            "  {} => {}",
            String::from_utf8_lossy(&key),
            String::from_utf8_lossy(&value)
        );
    }

    // 7. Durability -- ensure data is persisted to disk
    db.persist(PersistMode::SyncAll)?;

    println!("\nDone.");
    Ok(())
}
